pkg/operator/ceph/object/notification/provisioner.go

```diff
@@ -0,0 +1,167 @@
+// Note that the notifications are provisioned from the OBC's label controller
+// as there is no dedicated controller for bucket notifications
+package notification
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/coreos/pkg/capnslog"
+	bktv1alpha1 "github.com/kube-object-storage/lib-bucket-provisioner/pkg/apis/objectbucket.io/v1alpha1"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	cephclient "github.com/rook/rook/pkg/daemon/ceph/client"
+	"github.com/rook/rook/pkg/operator/ceph/object"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	awssession "github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/ceph/go-ceph/rgw/admin"
+	"github.com/pkg/errors"
+)
+
+const (
+	provisionerName = "ceph-bucket-notification-provisioner"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", provisionerName)
+
+type Provisioner struct {
+	Client      client.Client
+	Context     *clusterd.Context
+	ClusterInfo *cephclient.ClusterInfo
+	ClusterSpec *cephv1.ClusterSpec
+}
+
+func (p *Provisioner) getCephUser(username string, objStore *cephv1.CephObjectStore, objContext *object.Context) (accessKey string, secretKey string, err error) {
+	if len(username) == 0 {
+		err = errors.New("no user name provided")
+		return
+	}
+
+	// CephClusterSpec is needed for GetAdminOPSUserCredentials()
+	objContext.CephClusterSpec = *p.ClusterSpec
+	adminAccessKey, adminSecretKey, err := object.GetAdminOPSUserCredentials(objContext, &objStore.Spec)
+	if err != nil {
+		return
+	}
+
+	adminOpsClient, err := admin.New(objContext.Endpoint, adminAccessKey, adminSecretKey, &http.Client{})
+	if err != nil {
+		err = errors.Wrapf(err, "failed to build admin ops API connection to get user: %q", username)
+		return
+	}
+
+	var u admin.User
+	u, err = adminOpsClient.GetUser(context.TODO(), admin.User{ID: username})
+	if err != nil {
+		err = errors.Wrapf(err, "failed to get ceph user: %q", username)
+		return
+	}
+
+	logger.Infof("successfully fetched ceph user: %q", username)
+	accessKey = u.Keys[0].AccessKey
+	secretKey = u.Keys[0].SecretKey
+	return
+}
+
+func (p *Provisioner) createSession(namespace string, owner string) (*awssession.Session, error) {
+	objStores, err := p.Context.RookClientset.CephV1().CephObjectStores(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	// TODO: which store to take?
+	objStore := objStores.Items[0]
+
+	objContext, err := object.NewMultisiteContext(p.Context, p.ClusterInfo, &objStore)
+	if err != nil {
+		return nil, err
+	}
+
+	accessKey, secretKey, err := p.getCephUser(owner, &objStore, objContext)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO:
+	// * what is the difference between objContext.ZoneGroup and bucket.Spec.Endpoint.Region
```

Comment:
I would expect the `objContext` properties refer to the object store endpoint. I don't see a bucket.Spec in this section so maybe this is an old TODO?

Reply:
the `Create()` method, calling this `createSession()` method has a pointer to the OB, which has "endpoint" and "region" fields

Reply:
`bucket.Spec.Endpoint.Region` is just the placeholder for the was sdk for the region to use but has no real meaning for rgw AFAIK. Just for API compliance, I suppose. As for objContext.ZoneGroup is different, it represents the zone group the rgw process is part of.